conn.go

```diff
@@ -1430,6 +1529,34 @@ func (c *Conn) executeQuery(ctx context.Context, qry *Query) *Iter {
 	case *resultVoidFrame:
 		return &Iter{framer: framer}
 	case *resultRowsFrame:
+		if x.meta.newMetadataID != nil {
+			// Updating the result metadata id in prepared stmt
+			//
+			// If a RESULT/Rows message reports
+			//      changed resultset metadata with the Metadata_changed flag, the reported new
+			//      resultset metadata must be used in subsequent executions
+
+			stmtCacheKey := c.session.stmtsLRU.keyFor(c.host.HostID(), c.currentKeyspace, qry.stmt)
+			oldInflight, ok := c.session.stmtsLRU.get(stmtCacheKey)
+			if !ok {
+				// We didn't find the stmt in the cache, so we just re-prepare it
+				return c.executeQuery(ctx, qry)
+			}
+
+			newInflight := &inflightPrepare{
+				done: make(chan struct{}),
+				preparedStatment: &preparedStatment{
+					id:               oldInflight.preparedStatment.id,
+					resultMetadataID: x.meta.newMetadataID,
+					request:          oldInflight.preparedStatment.request,
+					response:         x.meta,
+				},
+			}
+
+			c.session.stmtsLRU.add(stmtCacheKey, newInflight)
+			return c.executeQuery(ctx, qry)
```

Comment:
So, we're expecting that when `<new_metadata_id>` is presented in **ROWS/RESULT** response the driver should update its cache to include the updated result metadata set and new `<result_metadata_id>` for subsequent executions and return the result. 
If `<new_metadata_id>` is presented in response, but the stmts cache doesn't contain the corresponding prepared stmt, the driver should return the result only.

Reply:
> So, we're expecting that when <new_metadata_id> is presented in ROWS/RESULT response the driver should update its cache to include the updated result metadata set and new <result_metadata_id> for subsequent executions and return the result.
If <new_metadata_id> is presented in response, but the stmts cache doesn't contain the corresponding prepared stmt, the driver should return the result only.

+1

Reply:
ok I found a way to trigger the `NEW_METADATA_ID` flag in the response but you basically have to add temporary code to make the driver NOT update the prepared entry when it receives an UNPREPARED, this is obviously not how it should work but it's a way to at least do some manual testing to make sure this flow is working properly.

In `Conn.executeQuery` replace the whole `case *RequestErrUnprepared` block with this:

```
case *RequestErrUnprepared:
		stmtCacheKey := c.session.stmtsLRU.keyFor(c.host.HostID(), c.currentKeyspace, qry.stmt)
		prep := &writePrepareFrame{
			statement: qry.stmt,
		}
		if c.version > protoVersion4 {
			prep.keyspace = qry.keyspace
		}

		framer, err := c.exec(c.ctx, prep, qry.trace)
		if err != nil {
			c.session.stmtsLRU.remove(stmtCacheKey)
			return &Iter{err: x, framer: framer}
		}

		_, err = framer.parseFrame()
		if err != nil {
			c.session.stmtsLRU.remove(stmtCacheKey)
			return &Iter{err: x, framer: framer}
		}
		return c.executeQuery(ctx, qry)
```
This will cause gocql to send an `EXECUTE` with the old result metadata ID even after the RE-PREPARE which will then cause C* to return `ROWS` with the new metadata.

Reply:
Awesome! I can test the whole metadata_changed flow with this one. Thanks a lot!

Reply:
I wrote an [integration test](https://github.com/worryg0d/gocql/commit/9717fa551bb05e1d088ce1defb3159dd5a299af5) for this mechanism. Could you please take a look at it to ensure that I didn't miss something?

Also, I figured out that the driver still uses old metadata when `skipMeta` is set. According to the specification, the driver should ignore it when the `Metadata_changed` flag is set, right?
```
            0x0008    Metadata_changed: if set, the No_metadata flag has to be unset
                      and <new_metadata_id> has to be supplied. This flag is to be
                      used to avoid a roundtrip in case of metadata changes for queries
                      that requested metadata to be skipped.
```

Reply:
The driver should prioritize the metadata from the response if it exists before using the metadata from the prepared statement object regardless of whether the request had the skip metadata flag set or not. 

Just to clarify that part of the spec that you quoted refers to the `NO_METADATA` flag on the *response* not the `SKIP_METADATA` flag on the *request*. The gocql driver always assumes the rows response doesn't have the metadata (i.e. `NO_METADATA` is set) when the `SKIP_METADATA` flag was set on the request but we should remove that assumption 👍

Reply:
The test isn't triggering the METADATA_CHANGED flag atm because of what I mentioned before but here's a way that it can be changed to trigger it: https://github.com/joao-r-reis/gocql/commit/bf39830147873c52e825c10f3855538778f3b92f 

One thing that is missing from the test though is checking result data integrity after the metadata change. Maybe inserting a row after the ALTER with data on both columns and then asserting that the data retrieved matches the values we just inserted.

Reply:
Awesome! I like the idea of changing the cache to force the driver to send queries with old metadata. Thank you!

> One thing that is missing from the test though is checking result data integrity after the metadata change
I added this to the test